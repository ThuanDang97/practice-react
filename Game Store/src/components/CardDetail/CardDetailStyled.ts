import { CardActions, CardContent, styled } from '@mui/material'

export const CardContentStyled = styled(CardContent)`
  max-height: 230px;
  overflow-y: scroll;
`

export const CardActionsStyled = styled(CardActions)`
  display: flex;
  justify-content: flex-end;
  color: ${({ theme }) => theme.palette.common.white};
`
