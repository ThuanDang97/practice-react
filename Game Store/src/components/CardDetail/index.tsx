import { useState } from 'react'
import {
  Card,
  CardContent,
  Collapse,
  IconButton,
  IconButtonProps,
  Link,
  styled,
  Typography,
  useTheme,
} from '@mui/material'
import ExpandMoreIcon from '@mui/icons-material/ExpandMore'
import {
  CardActionsStyled,
  CardContentStyled,
} from '@components/CardDetail/CardDetailStyled'

// Types
import { Game } from '@self-types/Game.types'

export interface ICardDetailProps {
  game: Game
}

interface ExpandMoreProps extends IconButtonProps {
  expand: boolean
}

const ExpandMore = styled((props: ExpandMoreProps) => {
  const theme = useTheme()

  return (
    <IconButton sx={{ color: theme.palette.background.paper }} {...props} />
  )
})(({ theme, expand }) => ({
  transform: !expand ? 'rotate(0deg)' : 'rotate(180deg)',
  transition: theme.transitions.create('transform', {
    duration: theme.transitions.duration.shortest,
  }),
}))

const CardDetail = ({ game }: ICardDetailProps) => {
  const {
    name,
    description_raw,
    released,
    platforms,
    genres,
    website,
    developers,
    publishers,
  } = game
  const theme = useTheme()
  const [expanded, setExpanded] = useState(false)

  const handleExpandClick = () => {
    setExpanded((prevValue) => !prevValue)
  }

  return (
    <Card
      sx={{
        maxWidth: 600,
        maxHeight: expanded ? 580 : 286,
        backgroundColor: theme.palette.grey[900],
        borderRadius: 3,
      }}
    >
      <CardContentStyled>
        <Typography
          color={theme.palette.common.white}
          fontWeight={700}
          fontSize={28}
          variant="h3"
          marginBottom={2}
        >
          About
        </Typography>
        {description_raw.split('###').map((desc, index) => (
          <Typography
            key={index}
            color={theme.palette.grey[400]}
            variant="body2"
          >
            {desc}
          </Typography>
        ))}
      </CardContentStyled>
      <CardActionsStyled
        sx={{ backgroundColor: theme.palette.grey[800] }}
        disableSpacing
      >
        {expanded ? 'Hide' : 'More'}
        <ExpandMore
          expand={expanded}
          onClick={handleExpandClick}
          aria-expanded={expanded}
          aria-label="show more"
        >
          <ExpandMoreIcon />
        </ExpandMore>
      </CardActionsStyled>
      <Collapse
        sx={{ backgroundColor: theme.palette.grey[800] }}
        in={expanded}
        timeout="auto"
        unmountOnExit
      >
        <CardContent>
          <Link
            sx={{ color: theme.palette.common.white }}
            fontWeight={700}
            href={website}
            underline="none"
          >
            {name}
          </Link>
          <Typography marginTop={1.6} color={theme.palette.grey[500]} paragraph>
            Released: {new Date(released).toLocaleDateString()}
          </Typography>
          <Typography color={theme.palette.grey[500]} paragraph>
            Platforms: {platforms.map((item) => item.platform.name).join(', ')}
          </Typography>
          <Typography color={theme.palette.grey[500]} paragraph>
            Genres: {genres.map((item) => item.name).join(', ')}
          </Typography>
          <Typography color={theme.palette.grey[500]} paragraph>
            Developers: {developers.map((item) => item.name).join(', ')}
          </Typography>
          <Typography color={theme.palette.grey[500]} paragraph>
            Publishers: {publishers.map((item) => item.name).join(', ')}
          </Typography>
        </CardContent>
      </Collapse>
    </Card>
  )
}

export default CardDetail
