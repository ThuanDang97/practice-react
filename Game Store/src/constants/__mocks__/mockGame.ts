export const MOCK_GAME_API = [
  {
    id: 3498,
    slug: 'grand-theft-auto-v',
    name: 'Grand Theft Auto V',
    ratings_count: 5927,
    description_raw: '',
    website: '',
    released: '2013-09-17',
    background_image:
      'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
    developers: [
      {
        name: '',
      },
    ],
    publishers: [
      {
        name: '',
      },
    ],
    parent_platforms: [
      {
        platform: {
          id: 186,
          name: 'Xbox Series S/X',
          slug: 'xbox-series-x',
        },
      },
    ],
    platforms: [
      {
        platform: {
          id: 1,
          name: 'PC',
          slug: 'pc',
        },
      },
    ],
    genres: [
      {
        name: 'Actions',
      },
    ],
    short_screenshots: [
      {
        id: -1,
        image:
          'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
      },
    ],
  },
]

export const MOCK_GAME_TRANSFORMED = [
  {
    id: 3498,
    slug: 'grand-theft-auto-v',
    name: 'Grand Theft Auto V',
    price: 19.99,
    ratings_count: 5927,
    description_raw: '',
    website: '',
    released: '2013-09-17',
    background_image:
      'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
    developers: [
      {
        name: '',
      },
    ],
    publishers: [
      {
        name: '',
      },
    ],
    parent_platforms: [
      {
        platform: {
          id: 186,
          name: 'Xbox Series S/X',
          slug: 'xbox-series-x',
        },
      },
    ],
    platforms: [
      {
        platform: {
          id: 1,
          name: 'PC',
          slug: 'pc',
        },
      },
    ],
    genres: [
      {
        name: 'Actions',
      },
    ],
    short_screenshots: [
      {
        id: -1,
        image:
          'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
      },
    ],
  },
]

export const MOCK_GAME_DATA = {
  id: 3498,
  slug: 'grand-theft-auto-v',
  name: 'Grand Theft Auto V',
  price: 19.99,
  ratings_count: 5927,
  description_raw: '',
  website: '',
  released: '2013-09-17',
  background_image:
    'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
  developers: [
    {
      name: '',
    },
  ],
  publishers: [
    {
      name: '',
    },
  ],
  parent_platforms: [
    {
      platform: {
        id: 186,
        name: 'Xbox Series S/X',
        slug: 'xbox-series-x',
      },
    },
  ],
  platforms: [
    {
      platform: {
        id: 1,
        name: 'PC',
        slug: 'pc',
      },
    },
  ],
  genres: [
    {
      name: 'Actions',
    },
  ],
  short_screenshots: [
    {
      id: -1,
      image:
        'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
    },
  ],
}
