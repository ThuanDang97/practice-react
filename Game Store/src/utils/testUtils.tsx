import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'
import { BrowserRouter as Router } from 'react-router-dom'

// Themes
import theme from '@self-types/Themes'

// Contexts
import { AuthProvider } from '@contexts/AuthProvider'
import { CartProvider } from '@contexts/CartProvider'

export const customRender = (component: ReactNode) => {
  return render(
    <ThemeProvider theme={theme}>
      <AuthProvider>
        <CartProvider>
          <Router>{component}</Router>
        </CartProvider>
      </AuthProvider>
    </ThemeProvider>,
  )
}

export * from '@testing-library/react'
export { customRender as render }
