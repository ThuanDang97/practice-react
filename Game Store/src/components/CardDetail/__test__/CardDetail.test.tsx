import { fireEvent, screen } from '@testing-library/react'

// Utils
import { render } from '@utils/testUtils'

// Components
import CardDetail, { ICardDetailProps } from '@components/CardDetail'

// Mocks
import { MOCK_GAME_DATA } from '@constants/__mocks__/mockGame'

// Types

describe('Card Detail component', () => {
  const props: ICardDetailProps = {
    game: MOCK_GAME_DATA,
  }
  it('should render Card Detail', () => {
    const { container } = render(<CardDetail {...props} />)

    expect(container).toMatchSnapshot()
  })

  it('should run toggle ', () => {
    render(<CardDetail {...props} />)

    const icon = screen.getByRole('button')

    fireEvent.click(icon)

    expect(screen.getByText(props.game.name)).toHaveTextContent(props.game.name)
  })
})
