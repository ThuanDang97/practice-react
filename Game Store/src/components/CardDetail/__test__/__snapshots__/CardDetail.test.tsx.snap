// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Detail component should render Card Detail 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1on6vww-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-u9l2rr-MuiCardContent-root"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-mpefaq-MuiTypography-root"
      >
        About
      </h3>
      <p
        class="MuiTypography-root MuiTypography-body2 css-k3bcd8-MuiTypography-root"
      />
    </div>
    <div
      class="MuiCardActions-root css-t35doo-MuiCardActions-root"
    >
      More
      <button
        aria-expanded="false"
        aria-label="show more"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-16o0grm-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
